### ISSUE TITLE
metrics.Accuracy is not calculated correctly when the first argument is of type float16

### ISSUE THREAD

**Issue by pgagarinov** [PyTorchLightning/pytorch-lightning#4840]
### To Reproduce
```
>a = tensor([0.5015, 0.5068, 0.4597, 0.5176, 0.5063, 0.4873, 0.5073, 0.5049, 0.4871,
        0.4939, 0.5132, 0.5151, 0.5269, 0.5229, 0.4797, 0.5435],
       device='cuda:0', dtype=torch.float16)
>b = tensor([1., 0., 1., 1., 1., 0., 1., 0., 0., 0., 0., 1., 1., 0., 0., 0.],
       device='cuda:0')

>acc = metrics.Accuracy(compute_on_step = True).to('cuda')

>acc(a, b)
tensor(0., device='cuda:0')
```

### Expected behavior
Same as for float32:
```
>a = tensor([0.5015, 0.5068, 0.4597, 0.5176, 0.5063, 0.4873, 0.5073, 0.5049, 0.4871,
        0.4939, 0.5132, 0.5151, 0.5269, 0.5229, 0.4797, 0.5435],
       device='cuda:0')
>b = tensor([1., 0., 1., 1., 1., 0., 1., 0., 0., 0., 0., 1., 1., 0., 0., 0.],
       device='cuda:0')

>acc(a, b)
tensor(0.6250, device='cuda:0')
```

### Additional context
I run training with precision=16, the model spits out float16 logits which I pass to F.softmax and then to metrics.Accuracy.
The type of F.softmax(logit) depends on whether precision=16 is specified or not. 
Precision, recall and F1 metrics seem to be calculated correctly.

**Comment by SkafteNicki** [comment]
The problem is this check:
https://github.com/PyTorchLightning/pytorch-lightning/blob/78076ea0d99e4ba1f76a7992b7090812258c0d4d/pytorch_lightning/metrics/utils.py#L98-L100
as it assumes `preds` are `float`. It will probably be solved by PR https://github.com/PyTorchLightning/pytorch-lightning/pull/4837 but need to check up on that.

**Comment by tchaton** [comment]
Hey @SkafteNicki,

Any update there ?

Best,
T.C

**Comment by SkafteNicki** [comment]
@tchaton just confirmed that it will be solved when PR #4837 is merged.
